## What is Data Wrangling?

- It is a process of cleaning the data by removing all the issues present, which makes it easier to process and analyse the data further.
- It is a process of gathering, collecting and transforming raw data into another format for better analysis, better understanding, better decision making, accessing in less time. Data Wrangling is also known as a Data Munging.

## Data Wrangling Python libraries

1. Pandas : Data Transformation and Data Analysis
2. NumPy : Numerical Anaysis
3. Matplotlib : Data Visualization
4. seaborn : Data Visualization
5. sklearn : Machine Learning

## To use pandas library we first have to install the library

- Syntax to install any libaray : !pip install library_name

In [1]:
!pip install pandas

Error processing line 3 of C:\Users\AMANT\anaconda3\lib\site-packages\googleapis_common_protos-1.56.4-py3.10-nspkg.pth:

  Traceback (most recent call last):
    File "C:\Users\AMANT\anaconda3\lib\site.py", line 169, in addpackage
      exec(line)
    File "<string>", line 1, in <module>
    File "<frozen importlib._bootstrap>", line 562, in module_from_spec
  AttributeError: 'NoneType' object has no attribute 'loader'

Remainder of file ignored


In [2]:
## install numpy

!pip install numpy

Error processing line 3 of C:\Users\AMANT\anaconda3\lib\site-packages\googleapis_common_protos-1.56.4-py3.10-nspkg.pth:

  Traceback (most recent call last):
    File "C:\Users\AMANT\anaconda3\lib\site.py", line 169, in addpackage
      exec(line)
    File "<string>", line 1, in <module>
    File "<frozen importlib._bootstrap>", line 562, in module_from_spec
  AttributeError: 'NoneType' object has no attribute 'loader'

Remainder of file ignored


## To get a list of all the libraries that are pre installed 

In [3]:
!pip list

Package                       Version
----------------------------- ---------------
absl-py                       1.4.0
alabaster                     0.7.12
anaconda-client               1.11.1
anaconda-navigator            2.4.0
and                           66.0.2
anyio                         3.5.0
appdirs                       1.4.4
argon2-cffi                   21.3.0
argon2-cffi-bindings          21.2.0
array-record                  0.2.0
arrow                         1.2.3
astroid                       2.14.2
astropy                       5.1
asttokens                     2.0.5
astunparse                    1.6.3
atomicwrites                  1.4.0
attrs                         22.1.0
Automat                       20.2.0
autopep8                      1.6.0
Babel                         2.11.0
backcall                      0.2.0
backports.functools-lru-cache 1.6.4
backports.tempfile            1.0
backports.weakref             1.0.post1
bcrypt                        3.2.0
beautif

Error processing line 3 of C:\Users\AMANT\anaconda3\lib\site-packages\googleapis_common_protos-1.56.4-py3.10-nspkg.pth:

  Traceback (most recent call last):
    File "C:\Users\AMANT\anaconda3\lib\site.py", line 169, in addpackage
      exec(line)
    File "<string>", line 1, in <module>
    File "<frozen importlib._bootstrap>", line 562, in module_from_spec
  AttributeError: 'NoneType' object has no attribute 'loader'

Remainder of file ignored


## You install a new library, and check version of a library

In [5]:
## Check pandas version

import pandas 
pandas.__version__

'1.5.3'

In [6]:
## Check the numpy version

import numpy 
numpy.__version__

'1.23.5'

## Updating a library or a package 
- Syntax : conda update --all (all the libraries will be updated)
- This is syntax has to be used only in anaconda prompt 

## How to use a library in jupyter notebook

In [8]:
import pandas
import numpy 

In [9]:
li = [10, 20, 30, 40]
li

[10, 20, 30, 40]

In [10]:
type(li)

list

In [11]:
## convert the above list into numpy array

numpy.array(li)

array([10, 20, 30, 40])

In [12]:
array(li)

NameError: name 'array' is not defined

In [13]:
numpy.array(li)

array([10, 20, 30, 40])

numpy.arange()
numpy.zeros()

In [15]:
import numpy as np
import pandas as pd

## here np and pd are  alias names given to numpy and pandas 

In [16]:
np.array(li)

array([10, 20, 30, 40])

In [17]:
dic = {
    'name' : ['Sam', 'John', 'Anil', 'Nayan'],
    'age' : [40, 32, 27, 57]
}
dic

{'name': ['Sam', 'John', 'Anil', 'Nayan'], 'age': [40, 32, 27, 57]}

In [18]:
type(dic)

dict

In [19]:
data = pd.DataFrame(dic)
data

name  age
0    Sam   40
1   John   32
2   Anil   27
3  Nayan   57

In [21]:
dir(pd)

['ArrowDtype',
 'BooleanDtype',
 'Categorical',
 'CategoricalDtype',
 'CategoricalIndex',
 'DataFrame',
 'DateOffset',
 'DatetimeIndex',
 'DatetimeTZDtype',
 'ExcelFile',
 'ExcelWriter',
 'Flags',
 'Float32Dtype',
 'Float64Dtype',
 'Float64Index',
 'Grouper',
 'HDFStore',
 'Index',
 'IndexSlice',
 'Int16Dtype',
 'Int32Dtype',
 'Int64Dtype',
 'Int64Index',
 'Int8Dtype',
 'Interval',
 'IntervalDtype',
 'IntervalIndex',
 'MultiIndex',
 'NA',
 'NaT',
 'NamedAgg',
 'Period',
 'PeriodDtype',
 'PeriodIndex',
 'RangeIndex',
 'Series',
 'SparseDtype',
 'StringDtype',
 'Timedelta',
 'TimedeltaIndex',
 'Timestamp',
 'UInt16Dtype',
 'UInt32Dtype',
 'UInt64Dtype',
 'UInt64Index',
 'UInt8Dtype',
 '__all__',
 '__builtins__',
 '__cached__',
 '__deprecated_num_index_names',
 '__dir__',
 '__doc__',
 '__docformat__',
 '__file__',
 '__getattr__',
 '__git_version__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_config',
 '_is_numpy_dev',
 '_libs',
 '_testing',
 '

## Libraries and their common alias names used by the Data Science Communities

In [23]:
## Data Analysis libraries
import pandas as pd
import numpy as np
 
## Data Visualization libraries
import matplotlib.pyplot as plt
import seaborn as sns

## Process of Data Wrangling

- Data Exploration
- Deal with missing values
- Reshape the data
- Filtering the data based on specific conditions
- If necessary, combine the multiple datasets
- Others 

## NumPy library

- NumPy - Numerical Python
- Mainly used for only numerical data anlysis
- Array is the core building block for NumPy


## NumPy Array

In [24]:
li 

[10, 20, 30, 40]

In [26]:
## convert the list into numpy array

a = np.array(li)
a

array([10, 20, 30, 40])

In [27]:
type(a)

numpy.ndarray

## Drawback of NumPy array

In [28]:
li = [10, 'Sam', 'Bombay', 45]
li

[10, 'Sam', 'Bombay', 45]

In [31]:
## Array can only store one data type

b = np.array(li)
b

array(['10', 'Sam', 'Bombay', '45'], dtype='<U11')

## Q. Why do we need a numpy array when we can work with lists

- Ans : Array's are more faster in processing compared to a list

In [32]:
import numpy as np
import time
import sys

In [33]:
l1 = range(100000)
l2 = range(100000)

In [36]:
a1 = np.arange(100000)
a2 = np.arange(100000)

In [38]:
## Add the elements of the list note the time to complete the operation

start = time.time()
result = [(x+y) for x, y in zip(l1, l2)]
print("python list took: ", (time.time() - start) * 1000)

python list took:  15.33818244934082


In [39]:
## Add the elements of both the array, note the time to complete the operation

start = time.time()
result = a1 + a2
print("numpy array took: ", (time.time() - start) * 1000)

numpy array took:  9.447336196899414


# NumPy Array's are preffered over python lists since we see that the processing is faster in a array

# NumPy Array Creation functions:

In [40]:
arr = np.arange(10)
arr

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [42]:
arr = np.arange(25)
arr

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
       17, 18, 19, 20, 21, 22, 23, 24])

In [45]:
ones = np.ones(15)
ones

array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.])

In [46]:
zeros = np.zeros(20)
zeros

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0.])

In [47]:
np.ones([3,2])

array([[1., 1.],
       [1., 1.],
       [1., 1.]])

In [48]:
np.zeros([3,3])

array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])

In [49]:
np.arange(5, 50)

array([ 5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21,
       22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38,
       39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49])

In [50]:
np.arange(5, 50, 3)

array([ 5,  8, 11, 14, 17, 20, 23, 26, 29, 32, 35, 38, 41, 44, 47])

In [51]:
np.arange(5, 51, 3)

array([ 5,  8, 11, 14, 17, 20, 23, 26, 29, 32, 35, 38, 41, 44, 47, 50])

In [53]:
np.linspace(10, 80, 5)    ## linspace stands for linear spacing

array([10. , 27.5, 45. , 62.5, 80. ])

##### np.linspace() is used to create an array with equal spacing between any two consecutive values of the array

- 27.5 - 10 = 17.5
- 45 - 27.5 = 17.5
- 62.5 - 45 = 17.5
- 80 - 62.5 = 17.5

In [54]:
np.linspace(10, 50, 5)

array([10., 20., 30., 40., 50.])

In [55]:
np.linspace(50, 250, 10)

array([ 50.        ,  72.22222222,  94.44444444, 116.66666667,
       138.88888889, 161.11111111, 183.33333333, 205.55555556,
       227.77777778, 250.        ])

In [56]:
np.random.randint(50, 150, 5)

array([136, 131, 133,  70, 130])

In [57]:
np.random.randint(50, 150, 5)

array([ 79, 117, 140, 127, 122])

In [61]:
np.random.rand(50)

array([0.351353  , 0.10787946, 0.70890169, 0.78326089, 0.54282113,
       0.17946063, 0.93531657, 0.51909311, 0.05534548, 0.43669711,
       0.27902619, 0.41929539, 0.02502946, 0.01921762, 0.09377778,
       0.75990687, 0.83688408, 0.46255629, 0.79905864, 0.75689036,
       0.32694713, 0.59559235, 0.96034463, 0.29936671, 0.13037814,
       0.83802625, 0.71625432, 0.78802395, 0.37958565, 0.91302503,
       0.95552873, 0.87772341, 0.05188521, 0.23185249, 0.01812762,
       0.78737063, 0.30808164, 0.29885696, 0.82451785, 0.58365302,
       0.32064099, 0.79628147, 0.73651791, 0.51581546, 0.67103936,
       0.58489534, 0.2861315 , 0.93213796, 0.54199872, 0.17104434])

In [62]:
np.full(5, 'Delhi')

array(['Delhi', 'Delhi', 'Delhi', 'Delhi', 'Delhi'], dtype='<U5')

In [63]:
np.full(8, "Dhoni")

array(['Dhoni', 'Dhoni', 'Dhoni', 'Dhoni', 'Dhoni', 'Dhoni', 'Dhoni',
       'Dhoni'], dtype='<U5')

In [64]:
np.full(10, 20)

array([20, 20, 20, 20, 20, 20, 20, 20, 20, 20])

In [65]:
np.full(20, 10)

array([10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10, 10,
       10, 10, 10])

In [66]:
np.full("Dhoni", 10)

TypeError: 'str' object cannot be interpreted as an integer